pkg/store/postings_codec.go

```diff
@@ -25,18 +28,184 @@ import (
 // significantly (to about 20% of original), snappy then halves it to ~10% of the original.
 
 const (
-	codecHeaderSnappy = "dvs" // As in "diff+varint+snappy".
+	codecHeaderSnappy         = "dvs" // As in "diff+varint+snappy".
+	codecHeaderStreamedSnappy = "dss" // As in "diffvarint+streamed snappy".
 )
 
+func decodePostings(input []byte) (closeablePostings, error) {
+	var df func([]byte) (closeablePostings, error)
+
+	switch {
+	case isDiffVarintSnappyEncodedPostings(input):
+		df = diffVarintSnappyDecode
+	case isDiffVarintSnappyStreamedEncodedPostings(input):
+		df = diffVarintSnappyStreamedDecode
+	default:
+		return nil, fmt.Errorf("unrecognize postings format")
+	}
+
+	return df(input)
+}
+
 // isDiffVarintSnappyEncodedPostings returns true, if input looks like it has been encoded by diff+varint+snappy codec.
 func isDiffVarintSnappyEncodedPostings(input []byte) bool {
 	return bytes.HasPrefix(input, []byte(codecHeaderSnappy))
 }
 
+// isDiffVarintSnappyStreamedEncodedPostings returns true, if input looks like it has been encoded by diff+varint+snappy streamed codec.
+func isDiffVarintSnappyStreamedEncodedPostings(input []byte) bool {
+	return bytes.HasPrefix(input, []byte(codecHeaderStreamedSnappy))
+}
+
+var snappyWriterPool, snappyReaderPool sync.Pool
+
+func estimateSnappyStreamSize(length int) int {
+	// Snappy stream writes data in chunks up to 65536 in size.
+	// The stream begins with bytes 0xff 0x06 0x00 0x00 's' 'N' 'a' 'P' 'p' 'Y'.
+	// Our encoded data also needs a header.
+	// Each encoded (or uncompressed) chunk needs tag (chunk type 1B + chunk len 3B) + checksum 4B.
+
+	// Mark for encoded data.
+	ret := len(codecHeaderStreamedSnappy)
+	// Magic snappy stream start.
+	ret += 10
+
+	const maxBlockSize = 65536
+
+	length = 5 * length / 4 // 1.25B per posting.
+
+	blocks := length / maxBlockSize
+
+	ret += blocks * snappy.MaxEncodedLen(maxBlockSize)
+	length -= blocks * maxBlockSize
+	if length > 0 {
+		ret += snappy.MaxEncodedLen(length)
+	}
+
+	return ret
+}
+
+func diffVarintSnappyStreamedEncode(p index.Postings, length int) ([]byte, error) {
+	out := make([]byte, 0, estimateSnappyStreamSize(length))
+	out = append(out, []byte(codecHeaderStreamedSnappy)...)
+	compressedBuf := bytes.NewBuffer(out[len(codecHeaderStreamedSnappy):])
+	var sw *snappy.Writer
+
+	uvarintEncodeBuf := make([]byte, binary.MaxVarintLen64)
+
+	pooledSW := snappyWriterPool.Get()
```

Comment:
Is there any way we could reuse the pool for grpc compression: https://github.com/thanos-io/thanos/blob/main/pkg/extgrpc/snappy/snappy.go?

Reply:
Added this in https://github.com/thanos-io/thanos/pull/6303/commits/4a037bdcae3164524902bc44705d821d3d42ac3a, perhaps it's not too bad to call the same pool across packages?

Reply:
I think the risk is that buffers for different use cases end up being vastly different in size, leading to some waste when resizing them. But I am not aware of any potential issues with correctness that could happen if we reuse the same pool from different packages.

Reply:
Snappy writer uses the same buffer sizes for each writer instance so this should be a good change in that regard 👍

Reply:
> @GiedriusS Can you help me understand the memory usage graph? Did the memory usage increase after deploying the change? Sorry it is not very clear when you deployed this change

No, it's consistently smaller now. Another example:

![image](https://user-images.githubusercontent.com/2377233/234367155-56ff7e57-52ef-4371-a84e-090a70db99af.png)
Change deployed around 14:00.